p2p/server.py

```diff
@@ -122,36 +143,121 @@ def __init__(self,
         )
 
         # Create and register peer in peer_pool
-        eth_peer = ETHPeer(
-            remote=initiator_remote, privkey=self.privkey, reader=reader,
-            writer=writer, aes_secret=aes_secret, mac_secret=mac_secret,
-            egress_mac=egress_mac, ingress_mac=ingress_mac, chaindb=self.chaindb,
+        peer = self.peer_class(
+            remote=initiator_remote,
+            privkey=self.privkey,
+            reader=reader,
+            writer=writer,
+            aes_secret=aes_secret,
+            mac_secret=mac_secret,
+            egress_mac=egress_mac,
+            ingress_mac=ingress_mac,
+            chaindb=self.chaindb,
             network_id=self.network_id
         )
-        self.peer_pool.add_peer(eth_peer)
-
-
-def decode_authentication(ciphertext: bytes,
-                          privkey: datatypes.PrivateKey
-                          ) -> Tuple[datatypes.PublicKey, bytes, datatypes.PublicKey]:
-    """
-    Decrypts and decodes the ciphertext msg.
-    Returns the initiator's ephemeral pubkey, nonce, and pubkey.
-    """
-    if len(ciphertext) < ENCRYPTED_AUTH_MSG_LEN:
-        raise ValueError("Auth msg too short: {}".format(len(ciphertext)))
-    elif len(ciphertext) == ENCRYPTED_AUTH_MSG_LEN:
-        sig, initiator_pubkey, initiator_nonce, _ = decode_auth_plain(
-            ciphertext, privkey)
+
+        await self.apply_p2p_handshake(peer)
+
+    async def apply_p2p_handshake(self, peer: BasePeer) -> None:
+        finish_p2p_handshake = False
+        try:
+            # P2P Handshake.
+            await wait_with_token(
+                peer.do_p2p_handshake(),
+                token=peer.cancel_token,
```

Comment:
This should use the server's cancel token. Basically, any `Server` methods that use `wait_with_token()` should use the server's token